startx.man
.\"
.\" Copyright 1993, 1998  The Open Group
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" Except as contained in this notice, the name of The Open Group shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from The Open Group.
.\"
.\"
.TH STARTX __appmansuffix__ __xorgversion__
.SH NAME
startx \- initialize an X session
.SH SYNOPSIS
.B startx
[ [
.I client
]
.I options
\&\.\|.\|. ] [
.B \-\^\-
[
.I server
] [
.I display
]
.I options
\&.\|.\|. ]
.SH DESCRIPTION
The \fBstartx\fP script is a front end to
.BR xinit (__appmansuffix__)
that provides a
somewhat nicer user interface for running a single session of the X
Window System.  It is often run with no arguments.
.PP
Arguments immediately following the
.I startx
command are used to start a client in the same manner as
.BR xinit (__appmansuffix__).
The special argument
.RB '--'
marks the end of client arguments and the beginning of server options.
It may be convenient to specify server options with startx to change on a
per-session basis the
default color depth, the server's notion of the number of dots-per-inch the
display device presents, or take advantage of a different server layout, as
permitted by the
.BR __XSERVERNAME__ (__appmansuffix__)
server and specified in the
.BR __XCONFIGFILEMAN__
configuration.  Some examples of specifying server arguments follow; consult
the manual page for your X server to determine which arguments are legal.
.RS
.PP
startx -- -depth 16
.PP
startx -- -dpi 100
.PP
startx -- -layout Multihead
.RE
#ifdef __SCOMAN__
.PP
To determine the client to run,
.B startx
looks for the following files, in order:
.RS
.PP
.I $(HOME)/.startxrc
.PP
.I __libdir__/sys.startxrc
.PP
.I $(HOME)/.xinitrc
.PP
.I __xinitdir__/xinitrc
.RE
.PP
#else
.PP
To determine the client to run,
.B startx
first looks for a file called
.I .xinitrc
in the user's home directory.  If that is not found, it uses
the file
.I xinitrc
in the
.I xinit
library directory.
#endif
If command line client options are given, they override this
behavior and revert to the
.BR xinit (__appmansuffix__)
behavior.
To determine the server to run,
.B startx
first looks for a file called
.I .xserverrc
in the user's home directory.  If that is not found, it uses
the file
.I xserverrc
in the
.I xinit
library directory.
If command line server options are given, they override this
behavior and revert to the
.BR xinit (__appmansuffix__)
behavior.  Users rarely need to provide a
.I .xserverrc
file.
See the
.BR xinit (__appmansuffix__)
manual page for more details on the arguments.
.PP
The system-wide
.I xinitrc
and
.I xserverrc
files are found in the
.I __xinitdir__
directory.
.PP
The
.I .xinitrc
is typically a shell script which starts many clients according to the
user's preference.  When this shell script exits,
.B startx
kills the server and performs any other session shutdown needed.
Most of the clients started by
.I .xinitrc
should be run in the background.  The last client should run in the
foreground; when it exits, the session will exit.  People often choose
a session manager, window manager, or \fIxterm\fP as the ''magic'' client.
.SH EXAMPLE
.PP
Below is a sample \fI\.xinitrc\fP that starts several applications and
leaves the window manager running as the ''last'' application.  Assuming that
the window manager has been configured properly, the user
then chooses the ''Exit'' menu item to shut down X.
.sp
.in +4
.nf
xrdb \-load $HOME/.Xresources
xsetroot \-solid gray &
xbiff \-geometry \-430+5 &
oclock \-geometry 75x75\-0\-0 &
xload \-geometry \-80\-0 &
xterm \-geometry +0+60 \-ls &
xterm \-geometry +0\-100 &
xconsole \-geometry \-0+0 \-fn 5x7 &
exec twm
.fi
.in -4
.SH "ENVIRONMENT VARIABLES"
.TP 25
DISPLAY
This variable gets set to the name of the display to which clients should
connect.  Note that this gets
.IR set ,
not read.
.TP 25
XAUTHORITY
This variable, if not already defined, gets set to
.IR $(HOME)/.Xauthority .
This is to prevent the X server, if not given the
.I \-auth
argument, from automatically setting up insecure host-based authentication
for the local host.  See the
.BR Xserver (__appmansuffix__)
and
.IR Xsecurity (__miscmansuffix__)
manual pages for more information on X client/server authentication.
.SH FILES
#ifdef __SCOMAN__
.TP 25
.I $(HOME)/.startxrc
Client to run.  Typically a shell script which runs many programs in
the background.
.TP 25
.I __libdir__/sys.startxrc
Client to use if the user has no
.I .startxrc
file.
#endif
.TP 25
.I $(HOME)/.xinitrc
Client to run.  Typically a shell script which runs many programs in
the background.
.TP 25
.I $(HOME)/.xserverrc
Server to run.  The default is
.IR X .
.TP 25
.I __xinitdir__/xinitrc
Client to run if the user has no
.I .xinitrc
file.
.TP 25
.I __xinitdir__/xserverrc
Server to run if the user has no
.I .xserverrc
file.
.SH "SEE ALSO"
.BR xinit (__appmansuffix__),
.BR X (__miscmansuffix__),
.BR Xserver (__appmansuffix__),
.BR __XSERVERNAME__ (__appmansuffix__),
.BR __XCONFIGFILEMAN__


ChangeLog
2006-05-27  Jeremy C. Reed  <reed@reedmedia.net>

	* startx.cpp:
	In comments, change contractions to be spelled out to fix
	cpp "missing terminating" errors with single quotes.

2006-04-30  Matthieu Herrb <matthieu.herrb@laas.fr>

	* configure.ac:
	Quote MCOOKIE to allow to specify a command with args (ie
	MCOOKIE="/usr/sbin/openssl rand -hex 16" ./configure ...)

2006-04-01  Adam Jackson  <ajax@freedesktop.org>

	* configure.ac:
	Bump to 1.0.2.

2006-02-25  Jeremy C. Reed  <reed@reedmedia.net>

	* startx.cpp:
	Bug #4449 <https://bugs.freedesktop.org/show_bug.cgi?id=4449>
	Be more portable (for different /bin/sh shells) for
	the arithmetic.

2005-12-20  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version for X11R7 release.

2005-12-14  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for final X11R7 release candidate.

2005-12-06  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* Makefile.am:
	Change *man_SOURCES ==> *man_PRE to fix autotools warnings.

2005-12-03  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Add m4 magic to allow AS_HELP_STRING to work with older versions
	of autoconf 2.57.

2005-12-03  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for X11R7 RC3 release.

2005-11-14  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* Makefile.am:
	* configure.ac:
	Make the programs used in the xinit scripts configurable.

2005-11-09  Alan Coopersmith  <alan.coopersmith@sun.com>

	* configure.ac: 
	Add settings for SHELL_CMD & ARCHMANDEFS to match monolith
	changes for bug #3180.

	AC_DEFINE(XORG,...) so it is listed as available server in xinit
	help output.

	* Makefile.am:
	Update cpp -D flags to match monolith changes for bug #3180.
	
2005-11-09  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for X11R7 RC2 release.

2005-10-18  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for RC1 release.

2005-10-16  Alan Coopersmith  <alan.coopersmith@sun.com>

	* configure.ac: 
	* Makefile.am: 
	Use cpp to substitute variables in man pages
	
2005-10-10  Alan Coopersmith  <alan.coopersmith@sun.com>

	* configure.ac: 
	* Makefile.am:
	Don't define HAS_COOKIE_MAKER if mcookie wasn't found.

2005-08-29  Daniel Stone  <daniel@freedesktop.org>

	Version 0.99.1.

	* configure.ac:
	* Makefile.am:
	Define HAS_COOKIE_MAKER and search for mcookie so startx will generate
	xauth tokens for the servers it starts.


startx.cpp
XCOMM!SHELL_CMD

XCOMM
XCOMM This is just a sample implementation of a slightly less primitive
XCOMM interface than xinit.  It looks for user .xinitrc and .xserverrc
XCOMM files, then system xinitrc and xserverrc files, else lets xinit choose
XCOMM its default.  The system xinitrc should probably do things like check
XCOMM for .Xresources files and merge them in, startup up a window manager,
XCOMM and pop a clock and serveral xterms.
XCOMM
XCOMM Site administrators are STRONGLY urged to write nicer versions.
XCOMM

unset DBUS_SESSION_BUS_ADDRESS
unset SESSION_MANAGER

#if defined(__SCO__) || defined(__UNIXWARE__) || defined(__APPLE__)

XCOMM Check for /usr/bin/X11 and BINDIR in the path, if not add them.
XCOMM This allows startx to be placed in a place like /usr/bin or /usr/local/bin
XCOMM and people may use X without changing their PATH.
XCOMM Note that we put our own bin directory at the front of the path, and
XCOMM the standard system path at the back, since if you are using the Xorg
XCOMM server theres a pretty good chance you want to bias the Xorg clients
XCOMM over the old system's clients.

XCOMM First our compiled path
bindir=__bindir__

case $PATH in
    *:$bindir | *:$bindir:* | $bindir:*) ;;
    *) PATH=$bindir:$PATH ;;
esac

XCOMM Now the "old" compiled path
#ifdef __APPLE__
oldbindir=/usr/X11R6/bin
#else
oldbindir=/usr/bin/X11
#endif

if [ -d "$oldbindir" ] ; then
    case $PATH in
        *:$oldbindir | *:$oldbindir:* | $oldbindir:*) ;;
        *) PATH=$PATH:$oldbindir ;;
    esac
fi

XCOMM Bourne shell does not automatically export modified environment variables
XCOMM so export the new PATH just in case the user changes the shell
export PATH
#endif

#if defined(__SCO__) || defined(__UNIXWARE__)
XCOMM Set up the XMERGE env var so that dos merge is happy under X

if [ -f /usr/lib/merge/xmergeset.sh ]; then
	. /usr/lib/merge/xmergeset.sh
elif [ -f /usr/lib/merge/console.disp ]; then
	XMERGE=`cat /usr/lib/merge/console.disp`
	export XMERGE
fi

userclientrc=$HOME/.startxrc
sysclientrc=LIBDIR/sys.startxrc
scouserclientrc=$HOME/.xinitrc
scosysclientrc=XINITDIR/xinitrc
#else
userclientrc=$HOME/.xinitrc
sysclientrc=XINITDIR/xinitrc
#endif

userserverrc=$HOME/.xserverrc
sysserverrc=XINITDIR/xserverrc
defaultclient=XTERM
defaultserver=XSERVER
defaultclientargs=""
defaultserverargs=""
defaultdisplay=":0"
clientargs=""
serverargs=""

#ifdef __APPLE__

if [ "x$X11_PREFS_DOMAIN" = x ] ; then
    X11_PREFS_DOMAIN=LAUNCHD_ID_PREFIX".X11"
fi

XCOMM Initialize defaults (this will cut down on "safe" error messages)
if ! defaults read $X11_PREFS_DOMAIN cache_fonts >& /dev/null ; then
    defaults write $X11_PREFS_DOMAIN cache_fonts -bool true
fi

if ! defaults read $X11_PREFS_DOMAIN no_auth >& /dev/null ; then
    defaults write $X11_PREFS_DOMAIN no_auth -bool false
fi

if ! defaults read $X11_PREFS_DOMAIN nolisten_tcp >& /dev/null ; then
    defaults write $X11_PREFS_DOMAIN nolisten_tcp -bool true
fi

XCOMM First, start caching fonts
if [ x`defaults read $X11_PREFS_DOMAIN cache_fonts` = x1 ] ; then
    if [ -x $bindir/font_cache ] ; then
        $bindir/font_cache &
    elif [ -x $bindir/font_cache.sh ] ; then
        $bindir/font_cache.sh &
    elif [ -x $bindir/fc-cache ] ; then
        $bindir/fc-cache &
    fi
fi

if [ -x XINITDIR/privileged_startx ] ; then
	# Don't push this into the background becasue it can cause
	# a race to create /tmp/.X11-unix
	XINITDIR/privileged_startx
fi

if [ x`defaults read $X11_PREFS_DOMAIN no_auth` = x0 ] ; then
    enable_xauth=1
else
    enable_xauth=0
fi

if [ x`defaults read $X11_PREFS_DOMAIN nolisten_tcp` = x1 ] ; then
    defaultserverargs="$defaultserverargs -nolisten tcp"
fi

if defaults read $X11_PREFS_DOMAIN dpi >& /dev/null ; then
    defaultserverargs="$defaultserverargs -dpi `defaults read $X11_PREFS_DOMAIN dpi`"
fi

for ((d=0; ; d++)) ; do
    [[ -e /tmp/.X$d-lock ]] || break
done
defaultdisplay=":$d"

#else
enable_xauth=1
#endif

#if defined(__SCO__) || defined(__UNIXWARE__)

XCOMM SCO -t option: do not start an X server
case $1 in
  -t)   if [ -n "$DISPLAY" ]; then
                REMOTE_SERVER=TRUE
                shift
        else
                echo "DISPLAY environment variable not set"
                exit 1
        fi
        ;;
esac
#endif

whoseargs="client"
while [ x"$1" != x ]; do
    case "$1" in
    XCOMM '' required to prevent cpp from treating "/*" as a C comment.
    /''*|\./''*)
	if [ "$whoseargs" = "client" ]; then
	    if [ x"$client" = x ] && [ x"$clientargs" = x ]; then
		client="$1"
	    else
		clientargs="$clientargs $1"
	    fi
	else
	    if [ x"$server" = x ] && [ x"$serverargs" = x ]; then
		server="$1"
	    else
		serverargs="$serverargs $1"
	    fi
	fi
	;;
    --)
	whoseargs="server"
	;;
    *)
	if [ "$whoseargs" = "client" ]; then
	    clientargs="$clientargs $1"
	else
	    XCOMM display must be the FIRST server argument
	    if [ x"$serverargs" = x ] && @@
		 expr "$1" : ':[0-9][0-9]*$' > /dev/null 2>&1; then
		display="$1"
	    else
		serverargs="$serverargs $1"
	    fi
	fi
	;;
    esac
    shift
done

XCOMM process client arguments
if [ x"$client" = x ]; then
    client=$defaultclient

    XCOMM if no client arguments either, use rc file instead
    if [ x"$clientargs" = x ]; then
        if [ -f "$userclientrc" ]; then
            client=$userclientrc
        elif [ -f "$sysclientrc" ]; then
            client=$sysclientrc
#if defined(__SCO__) || defined(__UNIXWARE__)
        elif [ -f "$scouserclientrc" ]; then
            client=$scouserclientrc
        elif [ -f "$scosysclientrc" ]; then
            client=$scosysclientrc
#endif
        fi

        clientargs=$defaultclientargs
    fi
fi

XCOMM process server arguments
if [ x"$server" = x ]; then
    server=$defaultserver

    XCOMM if no server arguments or display either, use defaults
    if [ x"$serverargs" = x -a x"$display" = x ]; then
	XCOMM For compatibility reasons, only use xserverrc if there were no server command line arguments
	if [ -f "$userserverrc" ]; then
	    server=$userserverrc
	elif [ -f "$sysserverrc" ]; then
	    server=$sysserverrc
	fi

	serverargs=$defaultserverargs
	display=$defaultdisplay
    fi
fi

if [ x"$enable_xauth" = x1 ] ; then
    if [ x"$XAUTHORITY" = x ]; then
        XAUTHORITY=$HOME/.Xauthority
        export XAUTHORITY
    fi

    removelist=

    XCOMM set up default Xauth info for this machine
    case `uname` in
    Linux*)
        if [ -z "`hostname --version 2>&1 | grep GNU`" ]; then
            hostname=`hostname -f`
        else
            hostname=`hostname`
        fi
        ;;
    *)
        hostname=`hostname`
        ;;
    esac

    authdisplay=${display:-:0}
#if defined(HAS_COOKIE_MAKER) && defined(MK_COOKIE)
    mcookie=`MK_COOKIE`
#else
    if [ -r /dev/urandom ]; then
        mcookie=`dd if=/dev/urandom bs=16 count=1 2>/dev/null | hexdump -e \\"%08x\\"`
    else
        mcookie=`dd if=/dev/random bs=16 count=1 2>/dev/null | hexdump -e \\"%08x\\"`
    fi
#endif
    if test x"$mcookie" = x; then
        echo "Couldn't create cookie"
        exit 1
    fi
    dummy=0

    XCOMM create a file with auth information for the server. ':0' is a dummy.
    xserverauthfile=$HOME/.serverauth.$$
    trap "rm -f '$xserverauthfile'" HUP INT QUIT ILL TRAP KILL BUS TERM
    xauth -q -f "$xserverauthfile" << EOF
add :$dummy . $mcookie
EOF
#if defined(__APPLE__) || defined(__CYGWIN__)
    serverargs=${serverargs}" -auth '"${xserverauthfile}"'"
#else
    serverargs=${serverargs}" -auth "${xserverauthfile}
#endif

    XCOMM now add the same credentials to the client authority file
    XCOMM if '$displayname' already exists do not overwrite it as another
    XCOMM server man need it. Add them to the '$xserverauthfile' instead.
    for displayname in $authdisplay $hostname$authdisplay; do
        authcookie=`XAUTH list "$displayname" @@
        | sed -n "s/.*$displayname[[:space:]*].*[[:space:]*]//p"` 2>/dev/null;
        if [ "z${authcookie}" = "z" ] ; then
            XAUTH -q << EOF 
add $displayname . $mcookie
EOF
        removelist="$displayname $removelist"
        else
            dummy=$(($dummy+1));
            XAUTH -q -f "$xserverauthfile" << EOF
add :$dummy . $authcookie
EOF
        fi
    done
fi

#if defined(__SCO__) || defined(__UNIXWARE__)
if [ "$REMOTE_SERVER" = "TRUE" ]; then
        exec SHELL_CMD ${client}
else
        XINIT "$client" $clientargs -- "$server" $display $serverargs
fi
#else

#if defined(__APPLE__) || defined(__CYGWIN__)
eval XINIT \"$client\" $clientargs -- \"$server\" $display $serverargs
#else
XINIT "$client" $clientargs -- "$server" $display $serverargs
#endif

#endif
retval=$?

if [ x"$enable_xauth" = x1 ] ; then
    if [ x"$removelist" != x ]; then
        XAUTH remove $removelist
    fi
    if [ x"$xserverauthfile" != x ]; then
        rm -f "$xserverauthfile"
    fi
fi

/*
 * various machines need special cleaning up
 */
#ifdef __linux__
if command -v deallocvt > /dev/null 2>&1; then
    deallocvt
fi
#endif

#ifdef macII
Xrepair
screenrestore
#endif

#if defined(sun)
kbd_mode -a
#endif

exit $retval

